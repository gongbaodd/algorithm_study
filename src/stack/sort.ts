import { Stack } from "./lib";

function moveGreaterBottomToTop(s: Stack<number>) {
  if (s.isEmpty()) {
    return;
  }

  const top1 = s.pop() as number;
  if (s.isEmpty()) {
    s.push(top1);
    return;
  }

  moveGreaterBottomToTop(s);
  const top2 = s.pop() as number;
  if (top1 < top2) {
    s.push(top1);
    s.push(top2);
  } else {
    s.push(top2);
    s.push(top1);
  }
}

export function sortStack(s: Stack<number>) {
  if (s.isEmpty()) {
    return;
  }

  moveGreaterBottomToTop(s);

  const top = s.pop() as number;
  sortStack(s);
  s.push(top);
}
