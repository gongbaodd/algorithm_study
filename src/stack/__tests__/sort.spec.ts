import { sortStack } from "../sort";
import { Stack } from "../lib";

describe("sort stack", () => {
  describe("provide an empty stack", () => {
    it("should do nothing", () => {
      const s = Stack.from([]);
      sortStack(s);

      expect(s.isEmpty()).toBe(true);
      expect(s).toMatchSnapshot();
    });
  });

  describe("provide a stack [2, 1, 4, 1, 5]", () => {
    it("should return 5,4,2,1,1", () => {
      const s = Stack.from([2, 1, 4, 1, 5]);
      sortStack(s);

      expect(s.isEmpty()).toBe(false);
      expect(s).toMatchSnapshot();
    });
  });
});
