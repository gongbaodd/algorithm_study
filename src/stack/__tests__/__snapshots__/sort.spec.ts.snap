// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sort stack provide a stack [2, 1, 4, 1, 5] should return 5,4,2,1,1 1`] = `
Stack {
  "pHead": LNode {
    "data": null,
    "next": LNode {
      "data": 5,
      "next": LNode {
        "data": 4,
        "next": LNode {
          "data": 2,
          "next": LNode {
            "data": 1,
            "next": LNode {
              "data": 1,
              "next": null,
            },
          },
        },
      },
    },
  },
}
`;

exports[`sort stack provide an empty stack should do nothing 1`] = `
Stack {
  "pHead": LNode {
    "data": null,
    "next": null,
  },
}
`;
